### PR TITLE
Benfrain pr/selector max specificity

### PR DESCRIPTION
**Pull request by davidtheclark**
I could not figure out a more elegant way with git to just contribute commits to @benfrain's PR (anybody know of one? @stylelint/core?)

Don't know if we want to close the original #754 or just add the differences from here to there.

My addition is pretty much just about `resolveNestedSelector()`, and the tests for it. I have a set of test-cases working with no `&`, initial `&` (direct nesting), and non-initial `&` (indirect nesting) --- though I didn't yet mess with the at-rule `@nest`.

What would be really helpful here would be contributions of more tests to ensure that `resolveNestedSelector()` works as expected. In fact, I think that that function would make a good standalone, thoroughly tested module (`postcss-resolve-nested-selector`??). If anybody is interested in picking up this code as-is and writing lots more tests, thinking through edge cases, and documenting, it would be a worthy OSS project :)

### CODE REVIEW

**Comment by ntwb** [comment]
I haven't done this for a while, and I think only once ever from memory:

You clone the fork and work directly in the same branch that the #754 pull request is using.

i.e. Clone `https://github.com/benfrain/stylelint/` and checkout the `pr/selector-max-specificity` branch, make your changes and push them back up to the same `pr/selector-max-specificity` branch at `benfrain/stylelint/`, they should then be included in the original pull request in #754. 

To make things easier, @benfrain could add you @davidtheclark as an owner/contributor to his fork so you can merge any changes directly to that fork/branch.

**Comment by benfrain** [comment]
If easier, as its effectively only 3 files just let me know the cleanest way to recreate the PR and I can redo?

**Comment by davidtheclark** [comment]
@ntwb: Yeah, essentially I'd be submitting PRs to his PR.

@benfrain: If you want to just copy over the files to your branch and create a new commit, it should work fine. (Only changed the test and the plugin.)

**Comment by davidtheclark** [comment]
I am going to try to create a module for resolve nested selectors. I should have a PoC in the next few days.

**Comment by benfrain** [comment]
That sounds great and I imagine it will prove useful for others too. Does it make sense for me to hang fire making any further commits until that's in place?

**Comment by davidtheclark** [comment]
Yeah I think so. I will work on it some tonight and try to get at least an
early version out before sleeping.

**Comment by davidtheclark** [comment]
@benfrain: There's a start here: https://github.com/davidtheclark/postcss-resolve-nested-selector

v0.1.0 is published on npm.

I tried it out and if you replace the function here with that one, the tests I wrote in still pass.

Anybody who wants to help develop that module let me know and I'll add you as a collaborator.

**Comment by benfrain** [comment]
Brilliant. Will add it in in about 10 hours and commit — excellent!

**Comment by benfrain** [comment]
The commit here includes the postcss-resolve-nested-selector work that @davidtheclark has done. Resolves the nested tests I had included. And also the ones added. Let me know if there are other specific patterns you want tests for and I will amend as needed.

File: src/rules/selector-max-specificity/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+.thing div {}
+````
+
+````
+.thing div {
+  a {}
+}
+````
+
+**Note**: currently [direct nesting](http://tabatkins.github.io/specs/css-nesting/#direct) is supported but the [@nest](http://tabatkins.github.io/specs/css-nesting/#at-nest) at-rule is not.
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/README.md]: Actually the `@nest` rule should be easy to support with the resolve-nested-selector module. The key is to look for at-rules, and if `theAtRule.name === 'nest'` split its `theAtRule.params` on `,` and use that as the selector array to check (like `rule.selectors`). The module should already do this when looking up the nesting chain (https://github.com/davidtheclark/postcss-resolve-nested-selector/blob/master/index.js#L3) and I also did it in the tests to specifically check how those `@nest` selectors resolve: https://github.com/davidtheclark/postcss-resolve-nested-selector/blob/master/test/util.js#L66

File: src/rules/selector-max-specificity/index.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            })
+            return
+          }
+        })
+      })
+    }
+  }
+}
+
+function resolveNestedSelector(selector, node) {
```
Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: Please replace this function with the postcss-resolve-nested-selector module

Comment (davidtheclark) [inline src/rules/selector-max-specificity/index.js]: By which I mean: install that module via npm as a dependency and import it.

**Comment by davidtheclark** [comment]
Regarding interpolation: I think we need to just run away when we see it. I did an experiment and see that in Sass I could create a variable `$foo: ul li;` and then interpolate it like `#{$foo} { top: 0; }` to get `ul li { top: 0; }`. You can do the same in Less. That means that we really can't check whether interpolation within a selector is or is not causing a specificity error. 

So here's what I think might work: if a selector matches this regular expression `/#{.+?}|@{.+?}|$\(.+?\)/` then that means it contains `#{..}` (Sass), or `@{..}` (Less), or `$(..)` (postcss-simple-vars) interpolation and we should ignore it. Sound ok?

**Comment by davidtheclark** [comment]
Closing in favor of #754

File: src/rules/selector-max-specificity/index.js (hunk 2 above)
Comment (benfrain) [inline src/rules/selector-max-specificity/index.js]: @davidtheclark are you looking at a different commit? I think I did this in https://github.com/benfrain/stylelint/commit/524040aa038e04ff214a766ece12f968126bf10f ?

File: src/rules/selector-max-specificity/README.md (hunk 1 above)
Comment (benfrain) [inline src/rules/selector-max-specificity/README.md]: Also, I removed this section of the README.md already in https://github.com/benfrain/stylelint/commit/524040aa038e04ff214a766ece12f968126bf10f

**Comment by davidtheclark** [comment]
@benfrain That commit you're referencing is not part of this PR, which is why I didn't see it. Check it out, and maybe if you can't find a solution with git open a new PR?

**Comment by benfrain** [comment]
Yes sorry. New to all this so probably made everything 10x more complicated than it should have been. 

Probably just easier if I open a completely new PR with all the latest in it. 

Is there anything missing in the commit I linked to or that you want changed in the new  PR?

**Comment by davidtheclark** [comment]
No problem! I screw up something git-related every day. That commit looks great. So I think the two outstanding pieces are

- Handling interpolation (run away)
- Handling `@nest` selectors

Sound right?
